### PR TITLE
Add Morpho USD Zap

### PR DESCRIPTION
**Pull request by Rubilmax**
A Curve Zap is usually built to ease deposits/withdrawals of assets (such as USDC, USDT, DAI) through Curve base pools & meta-pools. A base Curve pool is a generic stableswap AMM pool (e.g. DAI/USDC/USDT). A meta pool is a Curve pool on which you can deposit assets alongside base pool LP tokens (e.g. 3pool/LUSD).

`DepositZapMorphoAaveV2USD.vy` was built using a copy of `DepositZapUSD.vy`. However, changes were made: because there is no concept of meta-pools (yet) with the Morpho pool (it's just a simple maDAI/maUSDC/maUSDT pool), we don't bother allowing the user to deposit LP Tokens (and it doesn't even make sense).

So the template was much simplified because the user only needs to deposit/withdraw DAI/USDC/USDT through the zap, which should handle the logic of depositing/withdrawing on the underlying Morpho-Curve pool. The steps should be:
1. Deposit: transfer assets into the zap, deposit assets into Morpho vaults, then deposit maTokens into the Curve pool on behalf of the user
2. Withdraw: transfer Curve LP tokens into the zap, withdraw tokens from the Curve pool, redeem assets from Morpho vaults and set the user as receiver

Note: zaps don't need to comply to a specific interface, so I removed the `_pool` parameter used to pass the metapool address in the reference implementation (just as how Aave's curve zap is working: https://etherscan.io/address/0xDeBF20617708857ebe4F679508E7b7863a8A8EeE#code)

### CODE REVIEW

**Comment by MerlinEgalite** [comment]
Is there a codebase against which we can compare your code @Rubilmax ?

**Comment by Rubilmax** [comment]
> Is there a codebase against which we can compare your code @Rubilmax ?

Either the forked repository or `DepositZapUSD.vy` file (they are identical because this repo is a fork)
You can also review commit by commit to see the changes

**Comment by MerlinEgalite** [comment]
> > Is there a codebase against which we can compare your code @Rubilmax ?
> 
> Either the forked repository or `DepositZapUSD.vy` file (they are identical because this repo is a fork) You can also review commit by commit to see the changes

Ok I saw you added more info in the description thx!

File: contracts/zaps/DepositZapMorphoAaveV2USD.vy
Comment (MerlinEgalite) [inline contracts/zaps/DepositZapMorphoAaveV2USD.vy:60]: will it work with USDT?

File: contracts/zaps/DepositZapMorphoAaveV2USD.vy
Comment (MerlinEgalite) [inline contracts/zaps/DepositZapMorphoAaveV2USD.vy]: Should we pass `min_shares` or `min_amount` instead?

File: contracts/zaps/DepositZapMorphoAaveV2USD.vy
Comment (MerlinEgalite) [inline contracts/zaps/DepositZapMorphoAaveV2USD.vy]:
```suggestion
    _receiver: address = msg.sender
```

File: contracts/zaps/DepositZapMorphoAaveV2USD.vy
Comment (MerlinEgalite) [inline contracts/zaps/DepositZapMorphoAaveV2USD.vy]: it must be removed from the shares in my understanding of the code template

**Review by MerlinEgalite — CHANGES_REQUESTED** [review]
Some questions.

I need to spend more time on `remove_liquidity_imbalance`

File: contracts/zaps/DepositZapMorphoAaveV2USD.vy
Comment (Rubilmax) [reply contracts/zaps/DepositZapMorphoAaveV2USD.vy:60]: It seems that's what Curve is using in production... I'll double check again on multiple pools

File: contracts/zaps/DepositZapMorphoAaveV2USD.vy
Comment (Rubilmax) [reply contracts/zaps/DepositZapMorphoAaveV2USD.vy]: `min_shares` because the zap is withdrawing from the pool of maTokens (and `min_amount` is in assets)

File: contracts/zaps/DepositZapMorphoAaveV2USD.vy
Comment (pakim249CAL) [reply contracts/zaps/DepositZapMorphoAaveV2USD.vy:60]: I believe the assumption is that the only approval that the zapper will do is max(uint256) once for any token based on what I've seen in the other zappers. The BTC zapper keeps track of the approval per token per pool and approves the max if it hasn't been approved yet. This should remedy any issue with having to have allowance be zero before approval.

In regards to the return value, I don't think Vyper does the return value check, seeing as the interface defined does not define a return value for approve in the BTC zapper.

**Comment by pakim249CAL** [comment]
I'm considering whether this implementation really even fits the zapper paradigm that exists in the zapping template. We are really only wrapping from stablecoins to vault tokens, and then using that to call add liquidity to a pool, right? I think this doesn't really fit with what a zapper in this curve repo is supposed to do. 

The zappers in this repo exist to help zap into meta pools, not base pools. This is evidenced by the fact that the meta pool in the USD implementation is always called in a liquidity operation. It seems to have been built to help with the specific problem of zapping into the 3CRV + asset paradigm.

In our case, it may be nonsensical to fit into the interface of this paradigm. I think it'd be better to simply design a simple zapper from scratch instead of misleading people and developers into thinking that this is like a curve zapper.

File: contracts/zaps/DepositZapMorphoAaveV2USD.vy
Comment (Rubilmax) [reply contracts/zaps/DepositZapMorphoAaveV2USD.vy]: Removed the fee according to a dev from Curve: pure zaps should not take fees

**Review by pakim249CAL — APPROVED** [review]

**Review by MerlinEgalite — APPROVED** [review]

**Review by Tristan22400 — COMMENTED** [review]
looks good to me

**Comment by makcandrov** [comment]
LGTM

**Pull request merged**
